File: js-version/rough.js
class RoughPredictor {
	constructor(env) {
		this.env = env;
	}

	static predict(env, prng) {
		return new this(env).predict(prng);
	}

	predict(prng) {
		let [prngp, starters] = FactoryHelper.choose_entries(this.env, prng, this.env.nStarters);
		return this.predict0(prngp, [], [], starters);
	}

	predict0(prng, enemies, skipped, starters) {
		if (enemies.length == this.env.nBattles) {
			return [new RoughPredictorResult(prng, enemies, skipped, starters)];
		}
		let unchoosable = enemies.last || starters;
		let maybe_players = [...starters, ...enemies.slice(0, -1).flatten()];
		let results = OneEnemyPredictor.predict(env, prng, unchoosable, maybe_players);
		return results.map(result =>
			this.predict0(result.prng, [...enemies, result.chosen], [...skipped, result.skipped], starters)
		).flatten();
	}
}


class RoughPredictorResult {
	constructor(prng, enemies, skipped, starters) {
		this.prng = prng;
		this.enemies = enemies;
		this.skipped = skipped;
		this.starters = starters;
	}
}

class OneEnemyPredictor {
	constructor(env, unchoosable, maybe_players) {
		this.env = env;
		this.unchoosable = unchoosable;
		this.maybe_players = maybe_players;
	}

	static predict(env, prng, unchoosable, maybe_players) {
		return new this(env, unchoosable, maybe_players).predict(prng);
	}

	predict(prng) {
		return this.predict0(prng, [], [])
	}
	
	predict0(prng, skipped, chosen) {
		if (chosen.length == this.env.nParty) {
			return [new OneEnemyPredictorResult(prng, chosen, skipped)];
		}
		let [prngp, x] = FactoryHelper.choose_entry(this.env, prng);
		if (x.collides_within([...this.unchoosable, ...chosen, ...skipped])) {
			return this.predict0(prngp, skipped, chosen);
		} else if (!x.collides_within(this.maybe_players) || skipped.length == this.env.nParty) {
			return this.predict0(prngp, skipped, [...chosen, x]);
		} else {
			let result1 = this.predict0(prngp, skipped, [...chosen, x])
			let result2 = this.predict0(prngp, [...skipped, x], chosen)
			return [...result1, ...result2];
		}
	}
}

class OneEnemyPredictorResult {
	constructor(prng, chosen, skipped) {
		this.prng = prng;
		this.chosen = chosen;
		this.skipped = skipped;
	}
}

File: ruby-version/rough.rb
require "set"
require "pp"
require_relative "prng.rb"
require_relative "factory-helper.rb"

class RoughPredictor
  def initialize(env)
    @env = env
  end

  attr_reader :env
  include EnvMixin 
  include FactoryHelper

  def self.predict(env, prng)
    new(env).predict(prng)
  end

  def predict(prng)
    prng, starters = choose_entries(@env, prng, nStarters)
    predict0(prng, [], [], starters)
  end

  Result = Struct.new(:prng, :enemies, :skipped, :starters)

  def predict0(prng, enemies, skipped, starters)
    if enemies.length == nBattles
      return [Result.new(prng, enemies, skipped, starters)].to_set
    end
    unchoosable = enemies.last || starters
    maybe_players = starters + enemies[0..-2].flatten
    results = OneEnemyPredictor.predict(env, prng, unchoosable, maybe_players)
    results.map {|result|
      predict0(result.prng, enemies + [result.chosen], skipped + [result.skipped], starters)
    }.inject(:+)
  end
end

class OneEnemyPredictor
  def initialize(env, unchoosable, maybe_players)
    @env = env
    @unchoosable = unchoosable
    @maybe_players = maybe_players
  end

  attr_reader :env
  include EnvMixin 
  include FactoryHelper

  def self.predict(env, prng, unchoosable, maybe_players)
    new(env, unchoosable, maybe_players).predict(prng)
  end

  def predict(prng)
    predict0(prng, [], [])
  end

  Result = Struct.new(:prng, :chosen, :skipped)

  def predict0(prng, skipped, chosen)
    if chosen.length == nParty
      return [Result.new(prng, chosen, skipped)].to_set
    end
    prngp, x = choose_entry(@env, prng)
    if x.collides_within?(@unchoosable + chosen + skipped)
      predict0(prngp, skipped, chosen)
    elsif not x.collides_within?(@maybe_players) or skipped.length == nParty
      predict0(prngp, skipped, chosen + [x])
    else
      result1 = predict0(prngp, skipped, chosen + [x])
      result2 = predict0(prngp, skipped + [x], chosen)
      result1 + result2
    end
  end
end

if $0 == __FILE__
  require_relative "naive.rb"
  env = Env.new(nParty: 3, nStarters: 6, nBattles: 4, all_entries_file: "entries.csv")
  
  pp RoughPredictor.predict(env, PRNG.new(0)).map(&:enemies)
  exit

  10.times do
    seed = rand(2**32)
    print "%.8x: " % seed
    prng = PRNG.new(seed)
    result1 = RoughPredictor.predict(env, prng).map(&:enemies).to_set
    print "#{result1.size} results"
    result2 = NaivePredictor.predict(env, prng)
    print " / #{result2.size} results."
    puts " / #{result2.subset?(result1)}"
  end
end
